File: Assets/app/critical-strip/CriticalStripTransform.cs
using UnityEngine;
using UnityEngine.UIElements;

public class CriticalStripTransform
{
    private double minIndex;
    private double maxIndex;
    private double minImag; // New field for min imaginary value
    private double maxImag; // New field for max imaginary value
    private RectTransform viewportRect;
    private bool useImaginarySpace = false; // Default to index space

    // Cached rect values to avoid expensive RectTransform.get_rect() calls
    // The get_rect() method triggers C++ to C# marshalling which is very expensive
    // when called millions of times per frame during point rendering
    private float cachedRectWidth;
    private float cachedRectHeight;
    private float cachedRectX;
    private float cachedRectY;
    
    /// <summary>
    /// The critical line at real = 0.5 is of utmost importance in the Riemann hypothesis.
    /// All non-trivial zeros of the Riemann zeta function are conjectured to lie on this line.
    /// Due to floating-point arithmetic, transformations between viewport and strip coordinates
    /// might introduce tiny errors that could make points appear slightly off the critical line.
    /// 
    /// The threshold is dynamically calculated to be equivalent to 3 pixels on the current viewport width.
    /// This means any click within 3 pixels of the critical line will snap to 0.5.
    /// For example, with a 500px viewport, this creates a snap range of [0.497 to 0.503].
    /// 
    /// We err on the side of being more generous with the threshold because:
    /// 1. It's more important to correctly identify points intended to be on the critical line
    /// 2. Mouse precision and human hand steadiness typically have a margin of error of 2-3 pixels
    /// 3. The cost of a false positive (snapping a point that wasn't meant to be 0.5) is less
    ///    than a false negative (failing to snap a point that was meant to be 0.5)
    /// 
    /// You can use the "Analyze Critical Line Threshold" context menu item in PointSetManagerEditor
    /// to see exactly what range of values will be snapped to 0.5 with the current threshold.
    /// </summary>
    private const float CRITICAL_LINE_PIXELS = 3f; // Number of pixels to use for snapping
    
    public RectTransform ViewportRect => viewportRect;
    
    // Calculate threshold based on current viewport width
    public float CriticalValueThreshold => CRITICAL_LINE_PIXELS / cachedRectWidth;  // Use cached value
    
    // Property to get/set the space mode
    public bool UseImaginarySpace
    {
        get { return useImaginarySpace; }
        set 
        { 
            if (useImaginarySpace != value)
            {
                useImaginarySpace = value;
                // No need to update bounds here as they're already maintained in parallel
            }
        }
    }
    
    public CriticalStripTransform(RectTransform viewport, float minIndex = 0f, float maxIndex = 7f)
    {
        this.viewportRect = viewport;
        this.minIndex = minIndex;
        this.maxIndex = maxIndex;

        // Initialize imaginary values
        this.minImag = Zeta.IndexToImag(minIndex);
        this.maxImag = Zeta.IndexToImag(maxIndex);

        // Cache rect values to avoid expensive get_rect() calls
        InvalidateRectCache();

        // Debug.Log($"[CriticalStripTransform] Initialized with viewport width: {viewport.rect.width}, " +
        //           $"critical threshold: {CriticalValueThreshold:F6} ({CRITICAL_LINE_PIXELS} pixels)");
        // Debug.Log($"[CriticalStripTransform] Index range: [{minIndex}, {maxIndex}], Imag range: [{minImag:F2}, {maxImag:F2}]");
    }

    /// <summary>
    /// Invalidate the cached rect values and refresh from the RectTransform.
    /// Call this whenever the viewport size or position changes.
    /// </summary>
    public void InvalidateRectCache()
    {
        Rect rect = viewportRect.rect;
        cachedRectWidth = rect.width;
        cachedRectHeight = rect.height;
        cachedRectX = rect.x;
        cachedRectY = rect.y;
    }
    
    // Convert from critical strip coordinates (real [0,1], index/imag) to viewport coordinates
    public Vector2 StripToViewport(Vector2 stripPos)
    {
        int range = CriticalStripRenderer.realRange;
        float x;
        if (range == 0)
        {
            // real [0,1]
            x = stripPos.x * cachedRectWidth;  // Use cached value
        }
        else
        {
            // Shift the real axis so that 0.5 is at the center, then map [-realScale, realScale] to [0, cachedRectWidth]
            float shiftedX = stripPos.x - 0.5f;
            x = ((shiftedX + range) / (2f * range)) * cachedRectWidth;  // Use cached value
        }

        // Use double for calculations to maintain precision
        double normalizedY;
        if (useImaginarySpace)
        {
            normalizedY = (stripPos.y - minImag) / (maxImag - minImag);
            // Debug.Log($"[CriticalStripTransform] (imaginary) StripToViewport: {stripPos.y} -> {normalizedY}");
        }
        else
        {
            normalizedY = (stripPos.y - minIndex) / (maxIndex - minIndex);
            // Debug.Log($"[CriticalStripTransform] (index) StripToViewport: {stripPos.y} -> {normalizedY}");
        }
        float y = (float)(normalizedY * cachedRectHeight);  // Use cached value

        // Adjust for viewport position
        x += cachedRectX;  // Use cached value
        y += cachedRectY;  // Use cached value

        return new Vector2(x, y);
    }
    
    // Convert from viewport coordinates to critical strip coordinates
    public Vector2 ViewportToStrip(Vector2 viewportPos)
    {
        // Remove viewport position offset
        float adjustedX = viewportPos.x - cachedRectX;  // Use cached value
        float adjustedY = viewportPos.y - cachedRectY;  // Use cached value

        float normalizedX = adjustedX / cachedRectWidth;  // Use cached value

        // Check if we're close to the critical line (0.5)
        float distanceFromHalf = Mathf.Abs(normalizedX - 0.5f);
        float real;
        if (distanceFromHalf <= CriticalValueThreshold)
        {
            real = 0.5f;
        }
        else
        {
            float realMin = 0;
            float realMax = 1;
            // Support extended real axis range (e.g., -4.5 to 5.5)
            if (CriticalStripWindow.IsExtended)
            {
                realMin = -4.5f;
                realMax = 5.5f;
            }

            real = realMin + normalizedX * (realMax - realMin);
            real = Mathf.Clamp(real, realMin, realMax);
        }

        // Use double for y-coordinate calculations to maintain precision
        double normalizedY = adjustedY / cachedRectHeight;  // Use cached value
        double value;
        
        if (useImaginarySpace)
        {
            value = minImag + (normalizedY * (maxImag - minImag));
        }
        else
        {
            value = minIndex + (normalizedY * (maxIndex - minIndex));
        }
        
        return new Vector2(real, (float)value);
    }
    
    // Convert from screen coordinates to critical strip coordinates
    public Vector2 ScreenToStrip(Vector2 screenPos)
    {
        Vector2 localPos;
        bool success = RectTransformUtility.ScreenPointToLocalPointInRectangle(
            viewportRect, screenPos, null, out localPos);
            
        if (!success)
        {
            Debug.LogWarning($"[CriticalStripTransform] Failed to convert screen({screenPos}) to local coordinates");
            return Vector2.zero;
        }
        
        return ViewportToStrip(localPos);
    }
    
    // Set the range in the current coordinate space (index or imaginary)
    public void SetRange(float min, float max)
    {
        if (min >= max)
        {
            Debug.LogWarning($"[CriticalStripTransform] Invalid range: [{min}, {max}]");
            return;
        }
        
        if (useImaginarySpace)
        {
            minImag = min;
            maxImag = max;
            
            // Update corresponding index values
            minIndex = Zeta.ImagToIndex(minImag);
            maxIndex = Zeta.ImagToIndex(maxImag);
        }
        else
        {
            minIndex = min;
            maxIndex = max;
            
            // Update corresponding imaginary values
            minImag = Zeta.IndexToImag(minIndex);
            maxImag = Zeta.IndexToImag(maxIndex);
        }
        
        // Debug.Log($"[CriticalStripTransform] Range updated. Index: [{minIndex:F3}, {maxIndex:F3}], Imag: [{minImag:F2}, {maxImag:F2}]");
    }
    
    // Maintain backward compatibility
    public void SetIndexRange(float min, float max)
    {
        if (useImaginarySpace)
        {
            // If in imaginary space, convert the index values to imaginary values
            double minImag = Zeta.IndexToImag(min);
            double maxImag = Zeta.IndexToImag(max);
            SetRange((float)minImag, (float)maxImag);
        }
        else
        {
            SetRange(min, max);
        }
    }
    
    // Properties to get min/max values based on current space
    public float MinValue => (float)(useImaginarySpace ? minImag : minIndex);
    public float MaxValue => (float)(useImaginarySpace ? maxImag : maxIndex);
    
    // Keep original properties for backward compatibility
    public float MinIndex => (float)minIndex;
    public float MaxIndex => (float)maxIndex;
    
    // New properties to access imaginary range directly
    public float MinImag => (float)minImag;
    public float MaxImag => (float)maxImag;
} 

File: web/src/features/critical-strip/criticalStripTransform.ts
import { indexToImag } from "@/shared/math/zetaEms";
import type { SpaceMode, ViewRange } from "@/features/critical-strip/criticalStripTypes";

// 3 canvas-pixel snap zone for the critical line at σ=0.5
const CRITICAL_LINE_PIXELS = 3;

// Approximate inverse of indexToImag (from Zeta.cs ImagToIndex, Zzrob formula)
// eslint-disable-next-line no-loss-of-precision -- constant from mathematical reference, extra digits document the true value
const EULER_GAMMA = 0.5772156649015328606;
const E = Math.E;
const GAMMA_TO_E = Math.pow(EULER_GAMMA, E);       // ≈ 0.22451725198323
const TWO_ROOT_3_PI = 2 * Math.sqrt(3 * Math.PI);

export function imagToIndex(imag: number): number {
  return Math.sqrt(6 * GAMMA_TO_E / imag + 6 * imag + Math.PI) / TWO_ROOT_3_PI - 0.5;
}

export type CanvasRect = {
  width: number;
  height: number;
};

export type StripPos = { x: number; y: number };
export type ViewportPos = { x: number; y: number };

/**
 * Port of Unity CriticalStripTransform.cs.
 * Converts between strip space (σ, index/imag), viewport space (canvas px), and screen space.
 *
 * Strip X axis: σ mapped over sigmaRange (e.g. sigmaRange=1 → [0,1]; sigmaRange=5 → [-4,5])
 * Strip Y axis: index or imaginary t depending on spaceMode
 */
export class CriticalStripTransform {
  private minIndex: number;
  private maxIndex: number;
  private minImag: number;
  private maxImag: number;
  private spaceMode: SpaceMode;
  private sigmaRange: number;

  // Cached canvas dimensions — call invalidate() on resize
  private width = 1;
  private height = 1;

  public constructor(
    initialRange: ViewRange = { minY: 0, maxY: 7 },
    spaceMode: SpaceMode = "index",
    sigmaRange: number = 1,
  ) {
    this.spaceMode = spaceMode;
    this.sigmaRange = sigmaRange;

    this.minIndex = initialRange.minY;
    this.maxIndex = initialRange.maxY;
    this.minImag = indexToImag(this.minIndex, false);
    this.maxImag = indexToImag(this.maxIndex, false);
  }

  public invalidate(width: number, height: number): void {
    this.width = width;
    this.height = height;
  }

  public setSigmaRange(range: number): void {
    this.sigmaRange = range;
  }

  public setSpaceMode(mode: SpaceMode): void {
    this.spaceMode = mode;
  }

  public getSpaceMode(): SpaceMode {
    return this.spaceMode;
  }

  public setRange(minY: number, maxY: number): void {
    if (this.spaceMode === "imaginary") {
      this.minImag = minY;
      this.maxImag = maxY;
      this.minIndex = imagToIndex(minY);
      this.maxIndex = imagToIndex(maxY);
    } else {
      this.minIndex = minY;
      this.maxIndex = maxY;
      this.minImag = indexToImag(minY, false);
      this.maxImag = indexToImag(maxY, false);
    }
  }

  public getMinValue(): number {
    return this.spaceMode === "imaginary" ? this.minImag : this.minIndex;
  }

  public getMaxValue(): number {
    return this.spaceMode === "imaginary" ? this.maxImag : this.maxIndex;
  }

  public getMinIndex(): number { return this.minIndex; }
  public getMaxIndex(): number { return this.maxIndex; }
  public getMinImag(): number  { return this.minImag; }
  public getMaxImag(): number  { return this.maxImag; }

  // Critical-line snap threshold in strip-space units
  public get criticalValueThreshold(): number {
    return CRITICAL_LINE_PIXELS / this.width;
  }

  /**
   * Strip → canvas-pixel viewport coordinates.
   * Y convention: maxValue renders at y=0 (top), minValue at y=height (bottom).
   * Higher index values appear at the top of the canvas, matching Unity behavior.
   */
  public stripToViewport(strip: StripPos): ViewportPos {
    let x: number;
    if (this.sigmaRange === 1) {
      x = strip.x * this.width;
    } else {
      // maps [-(sigmaRange-1), sigmaRange] to [0, width]
      x = ((strip.x + (this.sigmaRange - 1)) / (2 * this.sigmaRange - 1)) * this.width;
    }

    const minY = this.spaceMode === "imaginary" ? this.minImag : this.minIndex;
    const maxY = this.spaceMode === "imaginary" ? this.maxImag : this.maxIndex;
    // normalizedY=0 → minValue (bottom), normalizedY=1 → maxValue (top)
    const normalizedY = (strip.y - minY) / (maxY - minY);
    // Y-down canvas: top=0, bottom=height → high values at y=0
    const y = (1 - normalizedY) * this.height;

    return { x, y };
  }

  /**
   * Canvas-pixel viewport coordinates → strip coordinates.
   * Inverse of stripToViewport.
   */
  public viewportToStrip(viewport: ViewportPos): StripPos {
    const normalizedX = viewport.x / this.width;

    let real: number;
    const distFromHalf = Math.abs(normalizedX - 0.5);
    if (this.sigmaRange === 1 && distFromHalf <= this.criticalValueThreshold) {
      real = 0.5;
    } else if (this.sigmaRange === 1) {
      real = normalizedX;
    } else {
      const realMin = -(this.sigmaRange - 1);
      const realMax = this.sigmaRange;
      real = realMin + normalizedX * (realMax - realMin);
      real = Math.max(realMin, Math.min(realMax, real));
      // snap near critical line in pixel space
      if (Math.abs(real - 0.5) < CRITICAL_LINE_PIXELS / this.width * (2 * this.sigmaRange - 1)) {
        real = 0.5;
      }
    }

    // Y-down: viewport.y=0 → maxValue (top), viewport.y=height → minValue (bottom)
    const minY = this.spaceMode === "imaginary" ? this.minImag : this.minIndex;
    const maxY = this.spaceMode === "imaginary" ? this.maxImag : this.maxIndex;
    const normalizedY = 1 - viewport.y / this.height;
    const value = minY + normalizedY * (maxY - minY);

    return { x: real, y: value };
  }

  /**
   * Convert a point stored as (real, index) in strip space to viewport coords.
   * Handles space-mode conversion automatically.
   */
  public pointToViewport(real: number, index: number): ViewportPos {
    const y = this.spaceMode === "imaginary" ? indexToImag(index, false) : index;
    return this.stripToViewport({ x: real, y });
  }

  /**
   * Convert imaginary-space viewport Y back to index.
   */
  public viewportYToIndex(viewportY: number): number {
    const minY = this.spaceMode === "imaginary" ? this.minImag : this.minIndex;
    const maxY = this.spaceMode === "imaginary" ? this.maxImag : this.maxIndex;
    // Y-down: y=0 → maxY, y=height → minY
    const normalizedY = 1 - viewportY / this.height;
    const value = minY + normalizedY * (maxY - minY);
    return this.spaceMode === "imaginary" ? imagToIndex(value) : value;
  }
}
